Backward and forward compatibility testing

thallgren opened the pull request:
This PR enables the integration tests to run with an traffic-manager/traffic-agent pair of a different version than the binary being tested, so that compatibility tests can be added for release candidates.

thallgren committed:
Move helm-specific helper functions to a separate itest/helm.go file.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Add ability to install Helm chart with version different from the client

The `telepresence helm install` and `telepresence helm upgrade` can now
handle a `--version` flag. When given, the corresponding Helm chart will
be downloaded from the "oci://ghcr.io/telepresenceio/telepresence-oss"
registry.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix backward compatibility issue when using replace

The new `telepresence replace` may use the port identifier "all" to
signify all ports of a targeted container, and this identifier is not
recognized by versions < 2.22.0 (no support for multi-port intercepts).

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Clean up the environment variable swamp related to integration testing.

This commit also enables backward compatibility testing by allowing
the traffic-manager and agent versions to be older than the version of
the tested binary.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix sporadic failure in argo-rollouts test.

The call to `telepresence uninstall` would sometimes race with pod
creation from the `telepresence intercept` call.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Prefer our own echo-server in favor of the jmalloc one.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Fix glitch causing namespace to not be deleted in CI test.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip commented on integration_test/multiple_services_test.go:
```diff
@@ -33,6 +33,9 @@ func init() {
 
 func (s *multipleServicesSuite) Test_LargeRequest() {
+	if !(s.ManagerIsVersion(">2.21.x") && s.ClientIsVersion(">2.21.x")) {
+		s.T().Skip("Not part of compatibility tests. TUN-device isn't stable enough versions <2.22.0")
```
```suggestion
		s.T().Skip("Not part of compatibility tests. TUN-device isn't stable enough in versions <2.22.0")
```
a small typo

P0lip commented on integration_test/multiple_services_test.go:
```diff
@@ -116,7 +119,7 @@ func (s *multipleServicesSuite) Test_LargeRequest() {
 	stdout := itest.TelepresenceOk(s.Context(), "-n", s.AppNamespace(), "list")
 	for i := 0; i < s.ServiceCount(); i++ {
-		s.Regexp(fmt.Sprintf(`%s-%d\s*: ready to engage`, s.Name(), i), stdout)
+		s.Regexp(fmt.Sprintf(`%s-%d\s*: ready to engage|intercept`, s.Name(), i), stdout)
```
```suggestion
		s.Regexp(fmt.Sprintf(`%s-%d\s*: ready to (engage|intercept)`, s.Name(), i), stdout)
```
nit, but the capturing (or non-capturing) group is needed here, otherwise the regexp matches a string `intercept`.
In other words, it's an alternative between:
- `%s-%d\s*: ready to engage`
- `intercept`

P0lip commented on integration_test/manual_agent_test.go:
```diff
@@ -109,7 +113,7 @@ func testManualAgent(s *itest.Suite, nsp itest.NamespacePair) {
 
 	stdout = itest.TelepresenceOk(ctx, "list")
-	require.Regexp(regexp.MustCompile(`.*`+ac.WorkloadName+`\s*:\s*ready to engage \(traffic-agent already installed\).*`), stdout)
+	require.Regexp(regexp.MustCompile(`.*`+ac.WorkloadName+`\s*:\s*ready to engage|intercept \(traffic-agent already installed\).*`), stdout)
```
```suggestion
	require.Regexp(regexp.MustCompile(`.*`+ac.WorkloadName+`\s*:\s*ready to (engage|intercept) \(traffic-agent already installed\).*`), stdout)
```

thallgren committed:
Update integration_test/multiple_services_test.go

Co-authored-by: Jakub Rożek <jrozek@datawire.io>
Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update integration_test/multiple_services_test.go

Co-authored-by: Jakub Rożek <jrozek@datawire.io>
Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Update integration_test/manual_agent_test.go

Co-authored-by: Jakub Rożek <jrozek@datawire.io>
Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Don't cache integration_test runs.

Let `go test` run with `-count=1` to disable caching. This is now
necessary since we run the `make check integration` target multiple
times with different environment settings to achieve the backward-
and forward compatibility tests.

Signed-off-by: Thomas Hallgren <thomas@tada.se>
